fastreplacestring/fastreplacestring.cpp

```diff
@@ -140,22 +141,48 @@ int main(int argc, char **argv) {
 
     // stat file so we can restore st_mode
     struct stat st;
-    stat(filename.c_str(), &st);
+    stat(filename, &st);
 
     // change st_mode so we can open file for writing
-    chmod(filename.c_str(), S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
-    in = freopen(filename.c_str(), "wb", in);
+    chmod(filename, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
+    in = freopen(filename, "wb", in);
     if (in == NULL) {
-      cout << "error: " + filename + " cannot be written to.\n";
+      free(t);
+      fclose(in);
+      fprintf(stderr, "error: %s cannot be written to.\n", filename);
       return 1;
     }
 
     fwrite(t, 1, newFilelen, in);
+    free(t);
     fclose(in);
 
     // restore st_mode
-    chmod(filename.c_str(), st.st_mode);
+    chmod(filename, st.st_mode);
   }
 
   return 0;
 }
+
+extern "C" {
+  CAMLprim value caml_fastreplacestring(value vPath, value vOldWord, value vNewWord) {
+    CAMLparam3(vPath, vOldWord, vNewWord);
+    CAMLlocal1(vRet);
+
+    char *szPath = String_val(vPath);
+    char *szOldWord = String_val(vOldWord);
+    char *szNewWord = String_val(vNewWord);
+
+    int ret = replace(szPath, szOldWord, szNewWord);
+    if (ret == 0) {
+      /* Ok() */
+      vRet = caml_alloc(1, 0);
+      Store_field (vRet, 0, Val_unit);
+    } else {
+      /* Error(..) */
+      vRet = caml_alloc(1, 1);
+      Store_field (vRet, 0, caml_copy_string("error rewriting file"));
```

Comment:
At first I was worried that we might hit a similiar issue to what you investigated with `mccs`, where we'd need to move the `caml_copy_string` out to a `CAMLlocal`, but I don't think that's necessary, since `Store_field` doesn't alloc.